proof/invariant-abstract/Tcb_AI.thy

```diff
@@ -1737,31 +1737,29 @@ lemma inj_on_snd_set_zip_map:
   "distinct xs \<Longrightarrow> inj_on snd (set (zip (map f xs) xs))"
   using distinct_map by fastforce
 
+lemma monadic_rewrite_is_valid':
+  "\<lbrakk> \<And>x. monadic_rewrite False True (P' x) (g x) (g' x); \<lbrace>P\<rbrace> f \<lbrace>P'\<rbrace>;
+    \<lbrace>P\<rbrace> do x <- f; g' x od \<lbrace>Q\<rbrace> \<rbrakk>
+   \<Longrightarrow> \<lbrace>P\<rbrace> do x <- f; g x od \<lbrace>Q\<rbrace>"
+  by (fastforce simp: monadic_rewrite_def valid_def bind_def)
```

Comment:
This looks a bit like monadic_rewrite_bind immediately applies and would yield precisely the kind of goals which you wrote as assumptions. Bit confused how this is working.

Reply:
I've removed this `monadic_rewrite` lemma and instead used `monadic_rewrite_refine_valid` and `monadic_rewrite_bind_tail` within the proof where the previous lemma was used.

Reply:
Cool, makes sense now